subsys/net/lib/dns/dns_sd.c

```diff
@@ -0,0 +1,881 @@
+	}
+
+	rsp->ancount++;
+	offset += r;
+
+	/* then add the additional records */
+	r = add_txt_record(inst, DNS_SD_TXT_TTL, instance_offset, buf,
+			   offset,
+			   buf_size - offset);
+	if (r < 0) {
+		return r; /* LCOV_EXCL_LINE */
+	}
+
+	rsp->arcount++;
+	offset += r;
+
+	r = add_srv_record(inst, DNS_SD_SRV_TTL, instance_offset,
+			   domain_offset,
+			   buf, offset, buf_size - offset, &host_offset);
+	if (r < 0) {
+		return r; /* LCOV_EXCL_LINE */
+	}
+
+	rsp->arcount++;
+	offset += r;
+
+	if (addr6 != NULL) {
+		r = add_aaaa_record(inst, DNS_SD_AAAA_TTL, host_offset,
+				    addr6->s6_addr,
+				    buf, offset,
+				    buf_size - offset); /* LCOV_EXCL_LINE */
+		if (r < 0) {
+			return r;                       /* LCOV_EXCL_LINE */
+		}
+
+		rsp->arcount++;
+		offset += r;
+	}
+
+	if (addr4 != NULL) {
+		tmp = htonl(*(addr4->s4_addr32));
+		r = add_a_record(inst, DNS_SD_A_TTL, host_offset,
+				 tmp, buf, offset,
+				 buf_size - offset);
+		if (r < 0) {
+			return r; /* LCOV_EXCL_LINE */
+		}
+
+		rsp->arcount++;
+		offset += r;
+	}
+
+	rsp->flags = htons(BIT(15) | BIT(10));
+	rsp->ancount = htons(rsp->ancount);
+	rsp->arcount = htons(rsp->arcount);
+
+	return offset;
+}
+
+/* TODO: dns_sd_handle_srv_query() */
+/* TODO: dns_sd_handle_txt_query() */
+
+bool dns_sd_rec_match(const struct dns_sd_rec *record,
+		      const struct dns_sd_rec *filter)
+{
+	size_t i;
+	const char *rec_label;
+	const char *filt_label;
+
+	static bool (*checkers[])(const char *) = {
+		instance_is_valid,
+		service_is_valid,
+		proto_is_valid,
+		domain_is_valid,
+	};
+
+	static const char *names[] = {
+		"instance",
+		"service",
+		"protocol",
+		"domain",
+	};
+
+	if (!rec_is_valid(record)) {
+		LOG_WRN("DNS SD record at %p is invalid", record);
+		return false;
+	}
+
+	if (filter == NULL) {
+		return false;
+	}
+
+	/* Deref only after it is deemed safe to do so */
+	const char *const pairs[] = {
+		record->instance, filter->instance,
+		record->service, filter->service,
+		record->proto, filter->proto,
+		record->domain, filter->domain,
+	};
+
+	BUILD_ASSERT(ARRAY_SIZE(pairs) == 2 * ARRAY_SIZE(checkers));
+	BUILD_ASSERT(ARRAY_SIZE(names) == ARRAY_SIZE(checkers));
+
+	for (i = 0; i < ARRAY_SIZE(checkers); ++i) {
+		rec_label = pairs[2 * i];
+		filt_label = pairs[2 * i + 1];
+
+		/* check for the "wildcard" pointer */
+		if (filt_label != NULL) {
+			if (!checkers[i](filt_label)) {
+				LOG_WRN("invalid %s label: '%s'",
+					names[i], filt_label);
+				return false;
+			}
+
+			if (strncasecmp(rec_label, filt_label,
+					DNS_LABEL_MAX_SIZE) != 0) {
+				return false;
+			}
+		}
+	}
+
+	/* check for the "wildcard" port */
+	if (filter->port != 0 && record->port != filter->port) {
+		return false;
+	}
+
+	return true;
+}
+
+int dns_sd_extract_service_proto_domain(const uint8_t *query,
+					size_t query_size, struct dns_sd_rec *record, char *service, size_t service_size,
+					char *proto, size_t proto_size, char *domain, size_t domain_size)
```

Comment:
Too long lines, try to limit the line length <80 chars

Reply:
Super weird - I ran this through `uncrustify` and that wacky alignment was the result.